crates/wasmparser/src/readers/core/elements.rs

```diff
@@ -283,10 +320,16 @@ impl<'a> ElementSectionReader<'a> {
             }
         }
         let data_end = self.reader.position;
-        let items = ElementItems {
-            offset: self.reader.original_offset + data_start,
-            data: &self.reader.buffer[data_start..data_end],
-            exprs,
+        let items = if exprs {
+            ElementItems::Expressions(ElementExpressionItemsReader::new(
+                &self.reader.buffer[data_start..data_end],
+                self.reader.original_offset + data_start,
+            )?)
+        } else {
+            ElementItems::Functions(ElementFunctionItemsReader::new(
+                &self.reader.buffer[data_start..data_end],
+                self.reader.original_offset + data_start,
+            )?)
```

Comment:
NB: this changes where the failure to parse the count of element items is reported. I’m happy to implement something that would allow detecting this particular error case separately as was possible before.